### ISSUE TITLE
[XLA] bazel tests broken

### ISSUE THREAD

**Issue by nouiz** [tensorflow/tensorflow#38205]
When running XLA C++ tests with bazel, many of them fails with an error like:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow/compiler/xla/client/lib:arithmetic_test_gpu
-----------------------------------------------------------------------------
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from ArithmeticTest
[ RUN      ] ArithmeticTest.ArgMinR2Axis0
2020-04-03 16:57:48.451927: I tensorflow/compiler/xla/service/platform_util.cc:72] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_gpu_jit)
2020-04-03 16:57:48.451981: F tensorflow/compiler/xla/tests/client_library_test_base.cc:48] Non-OK-status: result.status() status: Not found: no platforms found could not create local client for testing
external/bazel_tools/tools/test/test-setup.sh: line 310: 14802 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1
```

How to reproduce:

```
docker run --gpus all -it --rm tensorflow/tensorflow:devel-gpu /bin/bash
cd /tensorflow_src
./configure
git show
```
```
commit 09910a17749e7836462b843b537ffeccd302d260 (HEAD -> master, origin/master,
origin/HEAD)
Author: A. Unique TensorFlower <gardener@tensorflow.org>
Date:   Tue Mar 31 04:44:27 2020 -0700

    Fixed accumulator precision for generic DepthWise implementation.
    Removed inlined constants for kernel sizes.

    PiperOrigin-RevId: 303938922
    Change-Id: I4f1359341b8aa4a9c70b4c9f786657e5cbfdf645
```
```
bazel test -j 24 --config=opt -s  --config=noaws --config=nogcp --config=nohdfs --verbose_failures --config=cuda --cache_test_results=no --runs_per_test=1 --flaky_test_attempts=1 --test_tag_filters=gpu //tensorflow/compiler/xla/client/lib:arithmetic_test
cat /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/client/lib/arithmetic_test_gpu/test.log
```

See error above:

Now try with the current HEAD:

```
git pull
git show
```
```
commit f0571998d0195b5b243cf409a64d5fa17bd44d43 (HEAD -> master, origin/master, origin/HEAD)
Merge: 2921d2ed93 2082d706a5
Author: TensorFlower Gardener <gardener@tensorflow.org>
Date:   Fri Apr 3 09:29:53 2020 -0700

    Merge pull request #34218 from Flamefire:fix_missing_linker_path

    PiperOrigin-RevId: 304630859
    Change-Id: I3e412ed958acd0d60c8ddbdb22fd59ddf9caf05b
```
```
bazel test -j 24 --config=opt -s  --config=noaws --config=nogcp --config=nohdfs --verbose_failures --config=cuda --cache_test_results=no --runs_per_test=1 --flaky_test_attempts=1 --test_tag_filters=gpu //tensorflow/compiler/xla/client/lib:arithmetic_test
cat /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/client/lib/arithmetic_test_gpu/test.log
```

Same error.

I have this on 2 different computers.

**Comment by Saduf2019** [comment]
@nouiz 
as there is a PR to monitor this, please let us know if we may move this to closed status

**Comment by nouiz** [comment]
Which PR fix this? #38136 development is blocked by this issue.

**Comment by akuegel** [comment]
Would it work with the flag --noincompatible_do_not_split_linking_cmdline? We have seen this in some other tests as well that they only work when passing this flag to bazel.

**Comment by nouiz** [comment]
I tried with that flag with this command : `bazel test --config=nonccl --noincompatible_do_not_split_linking_cmdline --config=opt --config=cuda --cache_test_results=no //tensorflow/compiler/xla/client/lib:arithmetic_test`
And an extra flag with this command: `bazel test --config=nonccl --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 --noincompatible_do_not_split_linking_cmdline --config=opt --config=cuda --cache_test_results=no //tensorflow/compiler/xla/client/lib:arithmetic_test`

And in both cases, got this new error:
```
[ RUN      ] ArithmeticTest.ArgMinR2Axis0
2020-04-07 19:44:10.787375: I tensorflow/compiler/xla/service/platform_util.cc:72] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2020-04-07 19:44:10.787551: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-04-07 19:44:10.787585: E tensorflow/stream_executor/cuda/cuda_driver.cc:313] failed call to cuInit: UNKNOWN ERROR (-1)
2020-04-07 19:44:10.787609: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: a15e93e2e5b6
2020-04-07 19:44:10.787613: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: a15e93e2e5b6
2020-04-07 19:44:10.787706: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: Not found: was unable to find libcuda.so DSO loaded into this program
2020-04-07 19:44:10.787740: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 418.87.1
2020-04-07 19:44:10.787761: F tensorflow/compiler/xla/tests/client_library_test_base.cc:48] Non-OK-status: result.status() status: Not found: no CUDA devices found could not create local client for testing
external/bazel_tools/tools/test/test-setup.sh: line 310: 46321 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1
```

**Comment by cheshire** [comment]
I can reproduce the issue.

**Comment by cheshire** [comment]
Adding `linkstatic=1` to the `nvptx_compiler` target seems to fix the issue, but I'm trying to figure out what would be the unintended consequences.

**Comment by hlopko** [comment]
Hi all,

I've submitted a toolchain refactoring that makes the fix quite simple in https://github.com/tensorflow/tensorflow/commit/e0b19f6ef223af40e2e6d1d21b8464c1b2ebee8f. I could repro the failure mentioned above and I believe I understand how to best fix it (in short, moving `-no-as-needed` before libraries to link, and re-adding -lrt into nccl). I'll upload the fix next week.

**Comment by nouiz** [comment]
Great news! Thanks for continuing to work on this.

**Comment by nouiz** [comment]
For those following this issues, you will probably need to remote the `--noincompatible_do_not_split_linking_cmdline` parameter to bazel with this commit. This commit is already merged upstream. With it and the parameter, there is now a compilation error.

**Comment by nouiz** [comment]
I just tried it again with an updated devel-gpu container. 
With this command line:
`bazel test --config=opt -s  --config=noaws --config=nogcp --config=nohdfs --verbose_failures --config=cuda --cache_test_results=no --runs_per_test=1 --flaky_test_attempts=1 --test_tag_filters=gpu //tensorflow/compiler/xla/client/lib:arithmetic_test`

I have this error: 
```
Executing tests from //tensorflow/compiler/xla/client/lib:arithmetic_test_gpu
-----------------------------------------------------------------------------
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from ArithmeticTest
[ RUN      ] ArithmeticTest.ArgMinR2Axis0
2020-05-23 04:26:21.892802: I tensorflow/compiler/xla/service/platform_util.cc:72] platform CUDA present but no XLA compiler available: could not find registered compiler for platform CUDA -- check target linkage (hint: try linking in tensorflow/compile
r/jit:xla_gpu_jit)
2020-05-23 04:26:21.892842: F tensorflow/compiler/xla/tests/client_library_test_base.cc:48] Non-OK-status: result.status() status: Not found: no platforms found could not create local client for testing
external/bazel_tools/tools/test/test-setup.sh: line 310: 42559 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1
```

If I try with the `--noincompatible_do_not_split_linking_cmdline` parameter, bazel crash before running the test. The flag `--cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0` doesn't help more.

As said in a previous comment, your change remove the need for strange bazel flag at many places. This is pretty good. But there is still something missing to run the tests in a normal ubuntu container.

**Comment by hlopko** [comment]
Could you please verify that https://github.com/tensorflow/tensorflow/pull/39849 fixed the issue for you?

**Comment by nouiz** [comment]
I just tried and still have this error:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow/compiler/xla/client/lib:arithmetic_test_gpu
-----------------------------------------------------------------------------
2020-05-27 20:45:32.874308: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from ArithmeticTest
[ RUN      ] ArithmeticTest.ArgMinR2Axis0
2020-05-27 20:45:32.874883: I tensorflow/compiler/xla/service/platform_util.cc:72] platform Host present but no XLA compiler available: could not find registered compiler for platform Host -- check target linkage (hint: try linking in tensorflow/compiler/jit:xla_cpu_jit)
2020-05-27 20:45:32.875025: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-05-27 20:45:32.875039: E tensorflow/stream_executor/cuda/cuda_driver.cc:313] failed call to cuInit: UNKNOWN ERROR (-1)
2020-05-27 20:45:32.875063: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: a33883a7b4cb
2020-05-27 20:45:32.875068: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: a33883a7b4cb
2020-05-27 20:45:32.875226: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: Not found: was unable to find libcuda.so DSO loaded into this program
2020-05-27 20:45:32.875258: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 450.24.0
2020-05-27 20:45:32.875294: F tensorflow/compiler/xla/tests/client_library_test_base.cc:48] Non-OK-status: result.status() status: Not found: no CUDA devices found could not create local client for testing
external/bazel_tools/tools/test/test-setup.sh: line 310:  7224 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1
```

**Comment by hlopko** [comment]
I managed to repro and the "fix" is to remove stubs from `LD_LIBRARY_PATH`. I used 
```
cd /tensorflow_src
git pull
export LD_LIBRARY_PATH=/usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/include/x64_64-linux-gnu:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
bazel test -j 24 --config=noaws --config=nogcp --config=nohdfs --verbose_failures --config=cuda --cache_test_results=no --runs_per_test=1 --flaky_test_attempts=1 --test_tag_filters=gpu //tensorflow/compiler/xla/client/lib:arithmetic_test
```

and got a green run. Stubs were introduced in https://github.com/tensorflow/tensorflow/pull/30911. It seems that the PR expects that the stubs are overwritten but maybe they're not? @gunan do you have more context there?

**Comment by gunan** [comment]
Unfortunately, I do not have much context on that change.
@angerson may know. @yifeif in case she has ideas.

**Comment by angerson** [comment]
I'm not sure of how the stubs work, either. There's more context in #36974, where there are more suggestions for improvements.

**Comment by r4nt** [comment]
tl;dr: We did need stubs in the LD_LIBRARY_PATH back before we dlopened libcuda.

libcuda comes with the driver; if you don't have a GPU, you don't have a driver, and so you don't have libcuda. If you want to execute code that links against libcuda, you can make it point to the stub (which is basically "for linking when cross-compiling"). Executing code from stubs/libcuda will always return an error, but should otherwise be benign.

Given we now always dlopen libcuda on demand, we should not need the stub any more.

**Comment by r4nt** [comment]
(as an aside - if we need it for linking somewhere, the stubs LD_LIBRARY_PATH entry should always be after the "real" driver libcuda directory)

**Comment by nouiz** [comment]
> (as an aside - if we need it for linking somewhere, the stubs LD_LIBRARY_PATH entry should always be after the "real" driver libcuda directory)

This looks like a good solution.

@hlopko I tried with your suggested fix and I still had 4 tests failing. So it fix most of the problem, but not all of them. Here are the tests failing:
```
//tensorflow/compiler/xla/tests:local_client_execute_test_gpu            FAILED in 1 out of 30 in 34.9s
  Stats over 30 runs: max = 34.9s, min = 11.6s, avg = 16.0s, dev = 4.4s
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/local_client_execute_test_gpu/shard_28_of_30/test.log

//tensorflow/compiler/xla/tests:exhaustive_binary_16_bit_test_gpu        FAILED in 4 out of 48 in 45.3s
  Stats over 48 runs: max = 45.3s, min = 33.0s, avg = 36.3s, dev = 2.3s
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_16_bit_test_gpu/shard_1_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_16_bit_test_gpu/shard_31_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_16_bit_test_gpu/shard_17_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_16_bit_test_gpu/shard_15_of_48/test.log
//tensorflow/compiler/xla/tests:exhaustive_binary_test_f32_f64_gpu       FAILED in 4 out of 48 in 46.3s
  Stats over 48 runs: max = 46.3s, min = 32.1s, avg = 36.5s, dev = 3.1s
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_test_f32_f64_gpu/shard_26_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_test_f32_f64_gpu/shard_23_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_test_f32_f64_gpu/shard_15_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_binary_test_f32_f64_gpu/shard_12_of_48/test.log
//tensorflow/compiler/xla/tests:exhaustive_unary_test_complex_gpu        FAILED in 26 out of 48 in 82.7s
  Stats over 48 runs: max = 82.7s, min = 41.4s, avg = 62.7s, dev = 9.6s
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_unary_test_complex_gpu/shard_46_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_unary_test_complex_gpu/shard_1_of_48/test.log
  /root/.cache/bazel/_bazel_root/43801f1e35f242fb634ebbc6079cf6c5/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/compiler/xla/tests/exhaustive_unary_test_complex_gpu/shard_45_of_48/test.log
  ...
```

I checked 1 failure for each of them and they are related to nan not being present while they are expected. Example:
```
2020-06-01 14:39:05.700041: E tensorflow/compiler/xla/tests/exhaustive_op_test_utils.cc:314] Mismatch on (-nan (0xfdcb), -0 (0x8000)). Expected -nan (0xfe00), but got 1 (0x3c00).
```

Any guess on what could make it fail here but not for you? I ran this on a DGX1 with V100. On which hardware did you ran this container?

**Comment by akuegel** [comment]
The exhaustive_* tests are known to not be working in OSS, therefore they have a "no_oss" tag. They also don't work in the open source CI, only internally. I don't really know why, but maybe because they have shard_count > 1 and that causes problems in the open source environment? It seems local_client_execute_test also has shard_count > 1, but there are quite a few tests which do and which seem to work.

**Comment by nouiz** [comment]
I always forget about that tag. I guess it can be related to clang vs gcc, not about remote execution with the error. Maybe one of the compiler optimize the code differently and so do not handle nan exactly the same.

Sadly, if I'm right, it can mean that nan aren't well handle if not compiled with clang...

**Comment by nouiz** [comment]
The fix was included and a new container was made. Now it works.
Thanks.
